cmd/kubeadm/app/componentconfigs/kubelet.go

```diff
@@ -212,6 +216,65 @@ func (kc *kubeletConfig) Default(cfg *kubeadmapi.ClusterConfiguration, _ *kubead
 	}
 }
 
+// Mutate modifies absolute path fields in the KubeletConfiguration to be Windows compatible absolute paths.
+func (kc *kubeletConfig) Mutate() error {
+	// TODO: use build tags and move the Windows related logic to _windows.go files
+	// once the kubeadm code base is unit tested for Windows as part of CI - "GOOS=windows go test ...".
+	if runtime.GOOS != "windows" {
+		return nil
+	}
+
+	// When "kubeadm join" downloads the KubeletConfiguration from the cluster on Windows
+	// nodes, it would contain absolute paths that may lack drive letters, since the config
+	// could have been generated on a Linux control-plane node. On Windows the
+	// Golang path.IsAbs() function returns false unless the path contains a drive letter.
+	// This trips client-go and the kubelet, creating problems on Windows nodes.
+	// Fixing it in client-go or the kubelet is a breaking change to existing Windows
+	// users that rely on relative paths:
+	//   https://github.com/kubernetes/kubernetes/pull/77710#issuecomment-491989621
+	//
+	// Thus, a workaround here is to adapt the KubeletConfiguration paths for Windows.
+	// Note this is currently bound to KubeletConfiguration v1beta1.
+	klog.V(2).Infoln("[componentconfig] Adapting the paths in the KubeletConfiguration for Windows...")
+
+	// Get the drive from where the kubeadm binary was called.
+	exe, err := os.Executable()
+	if err != nil {
+		return errors.Wrap(err, "could not obtain information about the kubeadm executable")
+	}
+	drive := filepath.VolumeName(filepath.Dir(exe))
+	klog.V(2).Infof("[componentconfig] Assuming Windows drive %q", drive)
+
+	// Mutate the paths in the config.
+	mutatePathsOnWindows(&kc.config, drive)
+	return nil
+}
+
+func mutatePathsOnWindows(cfg *kubeletconfig.KubeletConfiguration, drive string) {
+	mutateStringField := func(name string, field *string, setEmpty bool) {
+		klog.V(2).Infof("[componentconfig] Adapting path for field %q with current value %q", name, *field)
+		// Set empty is a special case.
+		if setEmpty {
+			*field = ""
+			return
+		}
+		// path.IsAbs() is not reliable here in the Windows runtime, so check if the
+		// path starts with "/" instead. This means the path originated from a Unix node and
+		// is an absolute path.
+		if !strings.HasPrefix(*field, "/") {
+			return
+		}
+		// Prepend the drive letter to the path and update the field.
+		*field = filepath.Join(drive, *field)
+		return
+	}
+
+	// Mutate the fields we care about.
+	mutateStringField("resolvConf", cfg.ResolverConfig, true) // make sure this is disabled on Windows
+	mutateStringField("staticPodPath", &cfg.StaticPodPath, false)
+	mutateStringField("authentication.x509.clientCAFile", &cfg.Authentication.X509.ClientCAFile, false)
```

Comment:
This can be done only if the sub struct is not nil.

Reply:
I updated the code, and do a test according to https://gist.github.com/neolit123/1b7375c8a956155a2ca1cdd901336bce#file-test-mutate-kubeletconfig-windows-md

I probably got the correct result
![image](https://user-images.githubusercontent.com/13084946/141042711-2aae722f-930a-4961-a8a2-7e1e85964bf8.png)

Please check whether the judgment code I added is correct, thank you!

Reply:
so there was a problem that we cannot leave the ResolverConfig field to `nil`. otherwise it will get defaulted to this on Windows https://github.com/kubernetes/kubernetes/blob/656d00e8949f64cd663ca532d12e00982a6664c4/pkg/kubelet/types/constants.go#L21 https://github.com/kubernetes/kubernetes/blob/ef0eff14ab41f43b9f5f314b20489eb74911d414/pkg/kubelet/apis/config/v1beta1/defaults.go#L185 and it would fail on runtime.

https://github.com/kubernetes/kubernetes/blob/33de444861d3de783a6618be9d10fa84da1c11b4/pkg/kubelet/kubelet.go#L1995

this still seems like a problem, so we should do this change in mutatePathsOnWindows:

```go

import (

utilpointer "k8s.io/utils/pointer"
...

)

func mutatePathsOnWindows(cfg *kubeletconfig.KubeletConfiguration, drive string) {
	mutateStringField := func(name string)
		// path.IsAbs() is not reliable here in the Windows runtime, so check if the
		// path starts with "/" instead. This means the path originated from a Unix node and
		// is an absolute path.
		if !strings.HasPrefix(*field, "/") {
			return
		}
		// Prepend the drive letter to the path and update the field.
		*field = filepath.Join(drive, *field)
		klog.V(2).Infof("[componentconfig] kubelet/Windows: adapted path for field %q to %q", name, *field)
	}

	// Mutate the fields we care about.
	klog.V(2).Infof("[componentconfig] kubelet/Windows: adapting field \"resolverConfig\" to empty")
	if cfg.ResolverConfig == nil {
		cfg.ResolverConfig = utilpointer.String("") // Make sure this is empty on Windows
	}
	mutateStringField("staticPodPath", &cfg.StaticPodPath)
	mutateStringField("authentication.x509.clientCAFile", &cfg.Authentication.X509.ClientCAFile)
}
```

+ might need test update.